### PR TITLE
Vep preliminary implementation

### PR DESCRIPTION
**Pull request by lmoresi**
Waiting for the changes to constitutive models. In the meantime, I did some chasing down of the persistent memory leak. 

The biggest issues is in function.evaluate and occurs when we try to interpolate to swarm points in the mesh. I think the worst of this occurs in the c file, but I am not sure how accurate the simple print-based memory tracking is. Over to you @julesghub .

See the file Jupyterbook/Notebooks/Examples-StokesFlow/ISSUE_Function_Evaluation_mem_leak.py

to see the issue in action

### CODE REVIEW

**Comment by lmoresi** [comment]
Not for the purposes of merging, I'm making this so the memory leak issue is clear to @julesghub

**Comment by julesghub** [comment]
ok, thanks. I'll investigate now

**Comment by bknight1** [comment]
@lmoresi the changes I made to the VP/VEP constitutive models should be compatible with the old notebooks

**Comment by lmoresi** [comment]
> @lmoresi the changes I made to the VP/VEP constitutive models should be compatible with the old notebooks

OK - so that seems to be all good. Tests are passing after I merged all your changes back into my branch. I think we could merge my stuff into the BK branch and I can just work from that. Let me know when you are ready.

**Comment by lmoresi** [comment]
I have a question about the multi-material constitutive model @julesghub / @bknight1 

Is there a reason are we not just building a constitutive model per material and then masking by material when we construct the flux term. This would eliminate most of the code changes between the single material and the multi-material. It also fits better with the idea that the constitutive model is associated with a material and having a list of materials rather than having a large number of lists that have to be in sync all the time. That design pattern feels fragile to me. 

Am I missing something ?  (I expect I am)

L

**Comment by knepley** [comment]
That is what we do in PyLith. We mark sets of cells with a DMLabel, and then iterate over each set. The sets have different material models.

**Comment by lmoresi** [comment]
> 

@knepley, the issue we have is the particle-based materials. In PIC methods, we just used the particle material to build the constitutive law at each point. Using the PETSc point wise physics interface, we have adopted a strategy of applying level-set masks per material across the entire domain because we are required to supply the material variable in the form of a field. So this question is really about finding the least fragile way to present our approach to a typical user. 

Also, we are getting very close to a beta release. We should have a zoom with you in the loop to discuss details.

**Comment by knepley** [comment]
@lmoresi Oh, so what you want is a way for Swarm data to come in as the a[] array? We could do that instead of contortions. Would closest point to the quadrature point work?

**Comment by lmoresi** [comment]
> @lmoresi Oh, so what you want is a way for Swarm data to come in as the a[] array? We could do that instead of contortions. Would closest point to the quadrature point work?

We want that, yes. My question is not entirely related to that issue though. Since we are using the power of sympy for everything on the user end, we want to try to give sympy the chance to simplify / factorise expressions whenever we can. That leads to a slightly different view of how to express the problem and that's what I have been thinking about lately.

**Comment by julesghub** [comment]
> @lmoresi Oh, so what you want is a way for Swarm data to come in as the a[] array? We could do that instead of contortions. Would closest point to the quadrature point work?

@knepley Yes that would be fantastic to have a[] available at the point-wise evaluation. Let's make this happen!

**Comment by julesghub** [comment]
> I have a question about the multi-material constitutive model @julesghub / @bknight1
> 
> Is there a reason are we not just building a constitutive model per material and then masking by material when we construct the flux term. This would eliminate most of the code changes between the single material and the multi-material. It also fits better with the idea that the constitutive model is associated with a material and having a list of materials rather than having a large number of lists that have to be in sync all the time. That design pattern feels fragile to me.
> 
> Am I missing something ? (I expect I am)
> 
> L

We took the approach of one constitutive model per Model, rather than per material. 
Why? The "inspiration" was the idea of a single "kitchen sink" constitutive model that could collapse down to a given flow  with the power of sympy.
I do agree the current implementation is blotted dealing with single or multi-material.
Happy to discuss other ideas.

**Comment by bknight1** [comment]
> I have a question about the multi-material constitutive model @julesghub / @bknight1
> 
> Is there a reason are we not just building a constitutive model per material and then masking by material when we construct the flux term. This would eliminate most of the code changes between the single material and the multi-material. It also fits better with the idea that the constitutive model is associated with a material and having a list of materials rather than having a large number of lists that have to be in sync all the time. That design pattern feels fragile to me.
> 
> Am I missing something ? (I expect I am)
> 
> L

I was using a working implementation that would get the VP/VEP constitutive models working for now. A per-material approach would definitely reduce the bulk of the code, I think it would be worth discussing how to implement it.

**Comment by lmoresi** [comment]
I have had a crack at merging the MKMG (@knepley) with the JM stuff.

**Working** (perhaps):

Hierarchically refined meshes can now be used in the old solver pathway by John. The main problem here is that cloning a hierarchical dm is not a deep copy and the coarse dms have to be added carefully / linked correctly. Similar things are true when copying the various discrete systems and fields otherwise most of John's code falls over. That seems like a PETSc4py improvement on the horizon.

We still default to gamg with various tricks in the configuration, these can be overwritten to push geometrical mg (see SolCx example). This appears to work in serial and parallel but is only lightly tested. 

The auxiliary vectors are back so now the functions work properly for point wise functions / Jacobian evaluations and boundary conditions. 

The dm is not continually recreated / nuked during a solve but there is a bit of a convuluted pathway in the code because of the need to compile the boundary functions. 

I've got a feeling we can kill batman at some point ... 

**Not working**

No change to the meshVariable storage strategy.

Something is a bit weird with function.evaluate now (caching is breaking and maybe that's because we reset far less often).

We are stuck on the Knepley fixer-upper branch at the moment

No fixes to the D/Dt code that I promised to do. 

Pathways like FAS do not blow up any more but also don't seem to work (seems reasonable to assume that would be the case !!).

Everything to do with periodicity in parallel

**Questions/Observations**

In parallel, MG always says that it is just on one level even if the code is 5+ levels in serial. Is this possibly a PLEX setup problem ?

Is the deep clone of the dm hierarchy possible automatically or even desirable ?

Can we guarantee that a cloned DM will have the same distribution as the original. I broke that a few times but the current approach works on low numbers of cores.

**Comment by knepley** [comment]
1) Hmm, parallel MG should report the same number of levels, so something is going wrong. I suspect it is the timing of a distribution. We can debug it by viewing the mesh before each transformation and seeing where something goes wrong.

2) Not sure. I tend to be against it for memory use reasons.

3) Yes, a cloned DM has the same distribution.

**Comment by lmoresi** [comment]
OK - so I dug through and figured out the issues. The MG 1 level is actually a problem with the clone hierarchy that I messed up and that's fixed / works. The MG on multilevels is just fine now.

We have the usual problem that the interpolates for curved meshes produce points outside the grid if we move nodes to match the boundary. Fast if we don't. This is an interesting design choice that we need to make and I'd like to hear your thoughts, @knepley / @julesghub. 

  - Have an interpolated coordinate system $(r, \theta, \phi)$ but for $(x', y', z')$ shadowed on the mesh. This was Jed's preference as I recall but makes life harder for particle transport. Also, that's fine on the sphere, what about on topographic meshes ? Also this makes boundary conditions non-intuitive
  
  - Use a different interpolator (rbf doesn't care !) 
  - Stick with `gamg` since this doesn't care
  - Higher order meshes would make things less problematic (esp in 2d) 
  - Something else.

**Comment by knepley** [comment]
@lmoresi Do we have a representation of the boundary? If so, then this is easy. Refinement allows you to pass in a function that tells it how to alter the coordinates of any point it adds. You also get the labels for the point when you do this. So we could mark the boundary points, and it we get one, project the linear coordinates to the boundary. So far, this has worked well.

**Comment by lmoresi** [comment]
Ah - cool. I can’t see arguments so it might be a petsc4py fixer-upper. Or do we need to compile that ?

Sent from my iPhone

On 26 Aug 2023, at 10:31, Matthew Knepley ***@***.***> wrote:

﻿

@lmoresi<https://github.com/lmoresi> Do we have a representation of the boundary? If so, then this is easy. Refinement allows you to pass in a function that tells it how to alter the coordinates of any point it adds. You also get the labels for the point when you do this. So we could mark the boundary points, and it we get one, project the linear coordinates to the boundary. So far, this has worked well.

**Comment by knepley** [comment]
You are right. I need petsc4py scaffolding to allow a Python callback here. Hopefully you guys can show me how to do this.

**Comment by lmoresi** [comment]
> You are right. I need petsc4py scaffolding to allow a Python callback here. Hopefully you guys can show me how to do this.

Hi Matt, 

This is my solution which is, I think (with appropriate elegance and checking added), what petsc4py will end up doing:

```python

            self.dm_hierarchy = [self.dm]
            for i in range(refinement):
                dm_refined = self.dm_hierarchy[i].refine()
                dm_refined.setCoarseDM(self.dm_hierarchy[i])

                if callable(refinement_callback):
                    refinement_callback(dm_refined)

                self.dm_hierarchy.append(dm_refined)

```

The `refinement_callback` function is (in uw3) something the mesh object knows about because it is generally very specific to the geometry. 

There is also a callback to stuff particles back into the mesh if they fall out ... this is particularly relevant for curved boundaries and I think of it more like guard rails on a mountain road rather than reaching into the code  _deus ex machina_ 

Screen Shot 2023-09-02 at 7 21 45 am

*A screenshot of the annulus example ... the coarse mesh is an extreme example with two elements across the annulus so even with the callback, the refined mesh has a few funny artefacts. It does appear to do the right thing though*

**Comment by lmoresi** [comment]
Also, @knepley, the example above uses an approximate delta function to emulate a surface integral term that penalises radial flow - this is quite an effective way to add boundary conditions but we'd like to use the PETSc boundary integral paraphernalia to do this properly.

I have added the C code signatures for the point wise functions that are used in the various BdResidual / BdJacobian terms and now I am lost knowing how to light all those up. There seem to be no good examples to show what to do. Everything is either commented out or not quite using the machinery.

Is there an example somewhere of how this gets done ? 

I have these terms in the solver set up loop:

``` python
            PetscDSSetBdResidual(ds.ds, 0, 
                                ext.fns_bd_residual[i_bd_res[self._bd_f0]], ext.fns_bd_residual[i_bd_res[self._bd_f1]])

            ... # Jacobians / Preconditioned terms added as well
```

I then add a natural boundary condition (...NATURAL_FIELD) which needs a different point wise function added to it, labels and all that. But, as far as I can tell this is never activated. I can see it being identified when I look at the `dm.view()` though.

**Comment by knepley** [comment]
1) There are a few examples of using boundary integrals. SNES ex12 and ex71 do it.

2) Yes, that coordination between boundary conditions and boundary integrals is awkward. I need to make it nice now that the interface has settled down.

3) I made the code changes for global basis transformation today and mailed Julian. It is in knepley/feature-spherical-basis

**Comment by lmoresi** [comment]
I must admit the descriptions of the code and the implementation look nothing like the examples.

``` C
PetscErrorCode PetscDSSetBdResidual(PetscDS ds, PetscInt f, void (*f0)(PetscInt dim, PetscInt Nf, PetscInt NfAux, const PetscInt uOff[], const PetscInt uOff_x[], const PetscScalar u[], const PetscScalar u_t[], const PetscScalar u_x[], const PetscInt aOff[], const PetscInt aOff_x[], const PetscScalar a[], const PetscScalar a_t[], const PetscScalar a_x[], PetscReal t, const PetscReal x[], const PetscReal n[], PetscInt numConstants, const PetscScalar constants[], PetscScalar f0[]), void (*f1)(PetscInt dim, PetscInt Nf, PetscInt NfAux, const PetscInt uOff[], const PetscInt uOff_x[], const PetscScalar u[], const PetscScalar u_t[], const PetscScalar u_x[], const PetscInt aOff[], const PetscInt aOff_x[], const PetscScalar a[], const PetscScalar a_t[], const PetscScalar a_x[], PetscReal t, const PetscReal x[], const PetscReal n[], PetscInt numConstants, const PetscScalar constants[], PetscScalar f1[]))
{
  PetscFunctionBegin;
  PetscValidHeaderSpecific(ds, PETSCDS_CLASSID, 1);
  PetscCheck(f >= 0, PETSC_COMM_SELF, PETSC_ERR_ARG_OUTOFRANGE, "Field number %" PetscInt_FMT " must be non-negative", f);
  PetscCall(PetscWeakFormSetIndexBdResidual(ds->wf, NULL, 0, f, 0, 0, f0, 0, f1));
  PetscFunctionReturn(PETSC_SUCCESS);
}
```

compared to this in the examples:

``` c
    if (user->bcType == NEUMANN) {
      PetscCall(DMGetLabel(dm, "boundary", &label));
      PetscCall(DMAddBoundary(dm, DM_BC_NATURAL, "wall", label, 1, &id, 0, 0, NULL, NULL, NULL, user, &bd));
      PetscCall(PetscDSGetBoundary(ds, bd, &wf, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL));
      PetscCall(PetscWeakFormSetIndexBdResidual(wf, label, id, 0, 0, 0, f0_bd_u, 0, NULL));
    }

```

How does the boundary get detected in the first (documented) implementation ?

@knepley

**Comment by knepley** [comment]
1) The Swarm fixes you wanted are in https://gitlab.com/petsc/petsc/-/merge_requests/6856 and should merge soon. That has a top-level API for conservative projection and the Gaussian layout fix.

2) The `PetscDSSetBdResidual()` call is the old interface which is pretty limited, notably in that it assumes the problem has a single DS.

3) The new interface using DS is better. The call to `DMAddBoundary()` lets us know that we should integrate over this surface. The `PetscWeakFormSetIndexBdResidual()` sets a weak form with a key matching that boundary that should be integrated. Now that we have the WeakForm interface solid, you could argue that these calls should be combined, but I am not sure. The WF allows you to add many weak forms for that key, so it is very flexible. You could argue that the `DMAddBoundary()` call is redundant and could be reduced from the set of keys in the WF. Maybe we can pull that off.

**Comment by lmoresi** [comment]
Here is a comment on the latest bunch of refactoring

## Solver changes

I'm trying to align better the solver architecture with the PETSc approach. Specifically we now do this (in this order):
 
  1.  Identify all the required point-wise functions in a problem formulation, form appropriate derivatives and compile the whole collection at once (also make a dictionary using the functions as keys so they can be recovered whenever needed).
  2. Build the discretisation and add boundary conditions etc, load compiled point-wise functions into essential bcs
  3. Build the weak form engine by populating the Residuals, Jacobians etc with compiled functions.
  4. Setup the snes solver

Note: if the functions change, 1, 3 are revisited, 2 does not change, 4 is also triggered but may not need to be.

The code is in place to compile and load the boundary residuals / Jacobians as described above (but this does not currently light up in the solve - Needs fixing !). 

## Boundary conditions

Currently trying to simplify the bc's by forcing 1 component, 1 label, 1 label ID at a time. The scripts are a bit longer but the intent is much clearer and there is less confusion in loading this up from cython. 

## Hex meshes

Hex meshes seem to be working well now we have updated the required PETSc. CubedSphere now respects the refinement mechanisms and appears to work well for Q2/P1 and Q2/Q1 (added to tests).

## Bugs

  Something odd with regular simplex in 2D: boundary conditions not correctly applied on the domain corner (xfailing a test on this one)

## Miscellaneous

Fixed up many examples including the tests ... there are still some problems but no show-stoppers

**Comment by lmoresi** [comment]
No promises with this commit but:
  - running on the current PETSc main branch 2.19.5b 
  - Natural bc/s are mostly working
  - Rewriting the essential bc/s so that unconstrained components are flagged like this $(\infty, 1, \infty)$ using sympy as this is both compilable into a function that matches the requirement and easily detected to shuffle components into the constraints. This is what I broke previously. 
  - There is a bit of a mismatch with how I do natural bc/s so that does need some work.

I wrote some debug code for the point wise functions so that we can see when they are lit up and where. This is useful but only for very tiny meshes because it just prints to stdout.

  - Jacobians for non-linear natural bc/s are not lighting up properly and I can't see why that is. I am loading them into the same weak form as the bc/s themselves and I can see them there. Should that be done differently ?

Added a capture utility to grab those and other outputs so we can put them in tests.

**Comment by lmoresi** [comment]
A couple of fixes from Ben K's visit ... left-overs from changing the way the boundary conditions were applied. The natural bc/s and the relevant Jacobians are present and being called, but not exactly doing the right thing. There are some tests for these functions being compiled and lit up. The debugging code for the sympy compilation of point wise functions is more useful and the output is better controlled. 

I made a step towards killing off batman once and for all (he didn't see that coming did he ?). This is quite a simple change: when we add variables, we now check to see if the previous variables have been initialised and we copy the information into the resized vector. A bit clunky still ... and all the access manager code is still there (handles parallel sync, for example) but I suspect it's not all needed. There is no variable deletion at the moment but this would be a mirror image of the addition.

Also - adding a semi-Lagrange manager for DXDt which tries to match the fully-Lagrange manager (both need a refactor). This is almost done, but I didn't swap out the code in the solvers just yet ... to do this nicely really needs to have internal work-variables (hence the batman business above). 

L

**Comment by lmoresi** [comment]
Note: some merge race-conditions in this commit so this really needs checking

Trying to implement the batman-free Stokes solver to show how this should work (See SolCx ... ). The removal of the need to order everything allows us to create hidden data managers within solvers so that the user does not always need to know the data structures that should be internal to the computation (e.g. the swarm manager that SLCN uses to find launch points). 
This appears to be working OK. Tests are all passing. *NOTE* the messing about with the dm that holds the meshVariables is tested for non-hierarchical meshes but there may be some issues remaining when we use the mesh-refiners. 

Also implemented the semi-Lagrange and Lagrangian history manager classes so that they are interchangeable (I know that suggests there is a hierarchy that is not yet exploited). Using these in SLCN for Adv-diff and Navier-Stokes. The Lagrangian version of that is still a WIP.

**Comment by knepley** [comment]
Excellent. It sounds like we should have a meeting soon to get GMG worked out.

File: underworld3/systems/solvers.py [hunk 1]
```diff
@@ -107,6 +117,21 @@ def f(self, value):
...
         self._f = sympy.Matrix((value,))
 
+    @property
+    def constitutive_model(self):
+        return self._constitutive_model
+
+    @constitutive_model.setter
+    def constitutive_model(self, model):
+        if model is not None:
+            self._constitutive_model = model
```
Comment (julesghub) [inline underworld3/systems/solvers.py:116]: nice stuff, but shouldn't this be `self._constitutive_model = model(self.u)` and a straight replacement from what's in the __init__().

Comment (lmoresi) [reply underworld3/systems/solvers.py:116]: Ok - I'll bite. The init takes a class and the property is an instance. 

Should we give people the flexibility to call unexpected arguments for their models and pass them in ? 

Not sure ...  but I like the level of detail we are now discussing because it means we are much closer to finalising our schemas than we were a couple of months ago

Comment (lmoresi) [reply underworld3/systems/solvers.py:116]: I bit down on this prematurely and broke my teeth ... the above comment refers to my implementation but the new one from yesterday is different. Now Stokes can either have a fully implemented constitutive law or a class. Trouble is, if you pass a class, it's not working because we are no longer calling super.__init__(). I'm a bit confused here.

Comment (julesghub) [reply underworld3/systems/solvers.py:116]: Ideally both the class or implementation could be passed in and there'd be introspective function to checkout if they're full setup or not. I think that's the most flexible approach for now. 
Happy to discuss more at 2pm

**Comment by lmoresi** [comment]
I added some fixes so that the tests pass again.

I am not a fan of having arguments that can have different user-facing logic just determined by the type of the argument and so I went back to insisting that if you specify a class (rather than instance), you have to use a different argument name. Feel free to discuss !

The previous version removed most of the changes I had made to eliminate the Batman style (ie. Batman returns !) and I have tried to kick him back out. See SolCx for the example but obviously this needs a test of its own. 

Also see comments above - the tests were failing because we were muddling class and instance in the constitutive model setter. It a pattern that encourages errors so let's back away. 

Note to us: we probably need to put pressure in the Stokes constitutive model template. It might be needed soon ...

File: underworld3/constitutive_models.py
Comment (julesghub) [inline underworld3/constitutive_models.py:1]: Nice work - this is more flexible!

**Comment by lmoresi** [comment]
1) I can't get the tests to pass. Is this version of the code using some other functionality that I don't have ?

2) There is now a V_fn in Poisson which does not appear in any of the terms and we are defining all sorts of Lagrangian terms as a result. This seems to be the wrong way to name / construct the hierarchy. We could have Poisson be a specialised version of this class and do the same for Stokes / Navier-Stokes. Let's discuss this - I'd prefer the classes to be clean and easy to understand from the Jupyterlab strings. We don't want a new Matplotlib.

2a) We could alternatively define a Lagrangian derivative mix-in to add all the DuDt DFDt operations. Not sure I like the complexity but it would make the class definitions cleaner.

3) The constitutive model stuff is strictly correct, but I still do not like having this logic. I much prefer that we decide which way we think is the most effective and set the patterns accordingly. As far as I can tell, we can only use any constitutive model with one solver not pass it into multiple ones because of the way we connect unknowns to fluxes. We have unique unknowns for a given solver ... is that likely to change.

**Comment by julesghub** [comment]
1. I'm looking into now
2. Let's discuss this at 12. Sliding V_fn into Poison args was to harmonise all the solver constructors.
3. I don't know if it's likely to have other unknown connected to the fluxes. Can you think of a user case?

**Comment by lmoresi** [comment]
The tests failing occurs because the spherical model does not play well with Semi-Lagrange solvers. There can be small mismatches in coordinates and the result is that we can find the launch points in the swarm routines. 

We really should not be assuming that Stokes needs a Semi-Lagrange updater by default. We don't if we have a swarm updater.

The current pattern has disabled all the default constitutive model stuff but forces us to have all the building blocks for the constitutive models. Is this a merge issue on my part or is this a design choice ?

```
TypeError                                 Traceback (most recent call last)
Cell In[6], line 1
----> 1 stokes = uw.systems.Stokes(mesh, constitutive_model=uw.constitutive_models.ViscoElasticPlasticFlowModel)
      2 v = stokes._u
      3 p = stokes._p

TypeError: SNES_Stokes.__init__() got an unexpected keyword argument 'constitutive_model'
```

**Comment by julesghub** [comment]
1. Interesting about the Semi-Lagrangian not playing well with Spherical meshes. This makes sense. Our design choice was to endow all solvers with a vanilla Semi-lagrangian and a user can choose to overwrite it. We didn't consider misbehaving due to geometric situations.

2. The design choice is to remove constitutive arguments from the Solvers. You MUST set them later on.

**Comment by lmoresi** [comment]
1. We don't have good delete functionality for some of these more complicated objects so I think it is a mistake. Plus most people will forget to delete the DuDt / DFDt for vanilla Stokes. Would a better choice be to have a general template that includes all the terms and then specialised solvers that explicitly set those terms to be None ... then you would just need scalar / vector / saddle point type master solvers that correspond to a time-enabled version of the generic solver classes

1b. Swarm versions ... what is their status if we add Semi-Lagrange by default ?

2. Ah - OK. Explicitly construct the constitutive model after the solver has been initialised, not instantiating it inside the solver. 

Then we should also choose between this:

```python

stokes = uw.systems.Stokes(mesh)
v = stokes._u
p = stokes._p
stokes.constitutive_model = uw...VEFlowLaw(v, p)
stokes.constitutive_model.Parameters.shear_viscosity_0 = 1
```

and 

```python

stokes = uw.systems.Stokes(mesh)
stokes._u = uw.discretisation.MeshVariable("V1", ... )
stokes._p = uw.discretisation.MeshVariable("P1", ... )
stokes.constitutive_model = uw...VEFlowLaw(stokes._u , stokes._p)
stokes.constitutive_model.Parameters.shear_viscosity_0 = 1
```

**Comment by lmoresi** [comment]
Let's merge this into dev so we can start creating Fix-it-Felix branches
